File: tests_deps/src/main/java/com/github/klee0kai/test/car/di/bindinstance/simple/CarBindComponent.java
package com.github.klee0kai.test.car.di.bindinstance.simple;

import com.github.klee0kai.stone.annotations.component.Component;
import com.github.klee0kai.stone.annotations.module.BindInstance;
import com.github.klee0kai.stone.types.wrappers.Ref;
import com.github.klee0kai.test.car.model.Bumper;
import com.github.klee0kai.test.car.model.Wheel;
import com.github.klee0kai.test.car.model.Window;

import java.lang.ref.Reference;
import java.lang.ref.WeakReference;
import java.util.Collection;
import java.util.List;

@Component
public interface CarBindComponent {

    CarBindModule module();

    @BindInstance
    void bindWheel(Wheel wheel);

    @BindInstance
    void bindWheelRef(Reference<Wheel> wheel);

    @BindInstance
    void bindBumper(Ref<Bumper> bumper);

    @BindInstance
    void bindWindow(Window window);

    @BindInstance
    void bindWindows(Collection<Window> window);

    @BindInstance
    void bindWindowRefs(Collection<WeakReference<Window>> window);

    Wheel provideWheel();

    Reference<Wheel> provideWheelRef();

    List<Reference<Wheel>> provideWheels();

    Ref<Bumper> provideBumper();

    List<Bumper> provideBumpers();

    Window provideWindow();

    Collection<Window> provideWindows();

}

File: tests_wraps_kotlin/src/main/kotlin/com/github/klee0kai/test/car/di/bindinstance/simple/CarBindComponent.kt
package com.github.klee0kai.test.car.di.bindinstance.simple

import com.github.klee0kai.stone.annotations.component.Component
import com.github.klee0kai.stone.annotations.module.BindInstance
import com.github.klee0kai.stone.wrappers.Ref
import com.github.klee0kai.test.car.model.Bumper
import com.github.klee0kai.test.car.model.Wheel
import com.github.klee0kai.test.car.model.Window
import java.lang.ref.Reference
import java.lang.ref.WeakReference

@Component
interface CarBindComponent {
    fun module(): CarBindModule

    @BindInstance
    fun bindWheel(wheel: Wheel?)

    @BindInstance
    fun bindWheelRef(wheel: Reference<Wheel?>?)

    @BindInstance
    fun bindBumper(bumper: Ref<Bumper>?)

    @BindInstance
    fun bindWindow(window: Window?)

    @BindInstance
    fun bindWindows(window: Collection<Window?>)

    @BindInstance
    fun bindWindowRefs(window: Collection<WeakReference<Window>>)

    fun provideWheel(): Wheel
    fun provideWheelRef(): Reference<Wheel?>?
    fun provideWheels(): List<Reference<Wheel?>?>?
    fun provideBumper(): Ref<Bumper?>?
    fun provideBumpers(): List<Bumper?>?
    fun provideWindow(): Window
    fun provideWindows(): Collection<Window?>?
}
